### DISCUSSION TITLE
[help] Transient memory not working as expected

### DISCUSSION THREAD

**Discussion by Tristan-Siegfried** [ropensci/targets#1116] [Help]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Summary of Issue

I appear to get different memory clearing behavior (`memory = "transient", garbage_collection=T`) when I re-run the same targets with data generated by different mechanisms. I am confused about why the memory clearing is apparently being handled differently

### Reprex
```
library(targets)

tar_option_set(
  packages = "tidyverse",
  format = "qs",
  memory = "transient",
  garbage_collection = T,
  debug = "raw_data_mod"
)

list(
  tar_target(
    data_with_transcience_working,
    tibble(test=runif(8e7)) %>%
      {rep(list(.), 15)}
  ),
  tar_target(
    data_with_transcience_NOTworking,
    rep(list(diamonds), 2000)
  ),
  tar_target(
    raw_data_for_use_testing,
    # comment the first of these lines and un-comment the second of these for later step discussed below
    data_with_transcience_working
    #data_with_transcience_NOTworking,
  ),
  tar_target(
    testing_head,
    raw_data_for_use_testing %>% head(2)
  )
)
```

### Explanation of Reprex

1. First I run `tar_make(names=c("data_with_transcience_working", "data_with_transcience_NOTworking"))`. This sets up the data which will be used to later contrast the expected (working) vs unexpected (NOTworking) behavior of memory clearing. These targets are both just large lists of dataframes (can adjust the size if too large for your computer's memory), they just are created by different methods.
 
2. Then, with `data_with_transcience_working` un-commented in `raw_data_for_use_testing`, I run `tar_make()` and I get the following chart of memory usage:
![TransientSuccess](https://github.com/ropensci/targets/assets/129814507/b3daec8a-87fa-4bd8-872a-3516b7dacdc2)

This is as expected with `memory = "transient", garbage_collection=T`: I first see a large increase in memory usage as the pipeline is initialized and `raw_data_for_use_testing` is defined. Then (after a small increase which I don't totally understand but am guessing it is some overhead) I see a large decrease as `raw_data_for_use_testing` is cleared from memory, and then memory goes back up as `testing_head`  is generated as `raw_data_for_use_testing` needs to be reloaded. Then the memory usage goes down all the way as the pipeline finishes.

3. Then comment out `data_with_transcience_working` and un-comment `data_with_transcience_NOTworking` in the `raw_data_for_use_testing` target (thereby shifting the data to the `NOTworking` target), and lastly run `tar_make()`. I get the following chart of memory usage while running `tar_make()`:
![TransientFailure](https://github.com/ropensci/targets/assets/129814507/52a68f5c-16b3-4cd6-b434-a5fb611aaae3)

This is NOT as expected! Again, I see a large increase in memory usage as the pipeline is initialized and `raw_data_for_use_testing` is defined (this time to be equal to `data_with_transcience_NOTworking`), but then instead of memory being cleared before running `testing_head` I ONLY see memory usage goes up corresponding to running `testing_head`, with the pipeline finally finishing.

The issue is that in the first chart of memory usage I see an appropriate clearing of the `raw_data_for_use_testing` target from memory, but in the second chart I do not see `raw_data_for_use_testing` being cleared from memory before running `testing_head`.

### Session Info
Only used packages are `tidyverse` 2.0.0, and `targets` 1.2.2 on R 4.3.1.
Here is the full `sessionInfo`:
```
> sessionInfo()
R version 4.3.1 (2023-06-16)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: America/Chicago
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
 [1] targets_1.2.2   lubridate_1.9.2 forcats_1.0.0   stringr_1.5.0   dplyr_1.1.2     purrr_1.0.2     readr_2.1.4    
 [8] tidyr_1.3.0     tibble_3.2.1    ggplot2_3.4.2   tidyverse_2.0.0

loaded via a namespace (and not attached):
 [1] utf8_1.2.3        generics_0.1.3    renv_1.0.2        stringi_1.7.12    hms_1.1.3         digest_0.6.33    
 [7] magrittr_2.0.3    evaluate_0.21     grid_4.3.1        timechange_0.2.0  fastmap_1.1.1     processx_3.8.2   
[13] backports_1.4.1   ps_1.7.5          fansi_1.0.4       scales_1.2.1      codetools_0.2-19  cli_3.6.1        
[19] rlang_1.1.1       munsell_0.5.0     withr_2.5.0       yaml_2.3.7        tools_4.3.1       tzdb_0.4.0       
[25] colorspace_2.1-0  base64url_1.4     broom_1.0.5       vctrs_0.6.3       R6_2.5.1          lifecycle_1.0.3  
[31] pkgconfig_2.0.3   callr_3.7.3       pillar_1.9.0      gtable_0.3.3      glue_1.6.2        data.table_1.14.8
[37] xfun_0.40         tidyselect_1.2.0  rstudioapi_0.15.0 knitr_1.43        igraph_1.5.1      htmltools_0.5.6  
[43] rmarkdown_2.24    compiler_4.3.1 
```

**Comment by wlandau** [comment]
With the settings you describe, `targets` uses a combination of `rm()` and `gc()` to remove objects. The `diamonds` dataset is in the `ggplot2` package, and maybe being part of a package namespace has something to do with what you are seeing.

But to get a more accurate sense of what is happening, I recommend pausing the pipeline at each target you run for enough time for your system's memory monitoring to stabilize. The `debug` option of `tar_option_set()` could help, you could just set it to the names of all the targets, e.g. `tar_option_set(debug = c("data_with_transience_NOTworking", "raw_data_For_use_testing", "testing_head"))` (background: https://books.ropensci.org/targets/debugging.html).

It is also a bit less clear which targets are skipped and which are actually run in this case. E.g. `data_with_transience_NOTworking` and `raw_data_for_use_testing` were skipped, then the latter would only need to be loaded once.

Going forward, it would be great if these examples could be simplified into something more concise and targeted and easier to follow.

**Reply by Tristan-Siegfried** [reply]
> Going forward, it would be great if these examples could be simplified into something more concise and targeted and easier to follow.

Sorry about that, will try to be more targeted going forward!

Here is a smaller cleaned up reprex, no longer relying on any packages (other than `qs`) or pre-existing data. This also include the suggested debugging.

```
library(targets)

tar_option_set(
  format = "qs",
  memory = "transient",
  garbage_collection = T,
  debug = c("data_manual",
            "test_load1",
            "test_load2"
            )
)

list(
  tar_target(
    data_manual,
    data.frame(col1 = c("a", "b", "c"),
               col2 = c(1, 2, 3)) |>
      list() |>
      rep(x=_, 3e6)
  ),
  tar_target(
    test_load1,
    data_manual
  ),
  tar_target(
    test_load2,
    data_manual
    )
)
```

Here, when I go through and wait for memory to stabilize as I run the debugger I see the following chart of memory usage.
![Screenshot_20230821_214916](https://github.com/ropensci/targets/assets/129814507/fe3db03e-1fd3-4bce-af06-ef4aef0718ab)
Here there are 3 long flat sections (on the left, in the middle, on the right):
1. The first corresponds to the stabilized memory usage when the debugger prompted `The debugger is poised to run the command of target data_manual`. 
2. The second is the stabilized memory usage when the debugger prompted `The debugger is poised to run the command of target test_load1` (ie had already loaded `data_manual` into memory for use in that target) 
3. The third is the stabilized memory usage when the debugger prompted `The debugger is poised to run the command of target test_load2` (ie had already loaded `data_manual` into memory for use in that target) 

This is still not what I expected. I would expect that after I run `test_load1` but before I run `test_load2` that I would see `data_manual` being unloaded and then reloaded from memory when `targets` runs `rm()` and `gc()`. However, instead of releasing that memory and then re-loading the same object it appears memory usage just keeps climbing, almost to the point where it seems the object loaded into memory a full second time (without it ever being removed from memory) before running `test_load2`.

**Reply by wlandau** [reply]
The memory usage you see may also depend on the memory monitor you use. The `RES` column in output from `htop` is reliable, and I see far less usage there than on my MacBook's built-in Activity Monitor.

It seems like the `qs` package may be keeping its own copy of the data somehow. In an interactive debugger, when I call `unloadNamespace("qs")`, then memory usage halves. And when I run the pipeline with `format = "rds"` instead of `format = "qs"`, memory usage stays at around 2GB, which is about the size of `data_manual` in memory.

**Reply by wlandau** [reply]
Indeed, when I load data with `qs` and watch `RES` on `htop`, memory compounds on the first couple loads.

```r
packageVersion("qs")
#> [1] ‘0.25.5’

data <- data.frame(col1 = c("a", "b", "c"), col2 = c(1, 2, 3)) |>
  list() |>
  rep(x=_, 3e6)
tmp <- tempfile()
qs::qsave(data, tmp)

# RES 108M

loaded <- qs::qread(tmp) # 1.73 GB
gc()

# RES 1790M

loaded <- qs::qread(tmp)
gc()

# RES 3499M
```

**Reply by wlandau** [reply]
Submitted https://github.com/traversc/qs/issues/85

**Comment by traversc** [comment]
I think it's worth a brief followup of https://github.com/traversc/qs/issues/85

Consider this example in base R
```
data <- data.frame(col1 = c("a", "b", "c"), col2 = c(1, 2, 3)) |>
  list() |>
  rep(x=_, 3e6)
tmp <- tempfile()
saveRDS(data, tmp)

loaded <- readRDS(tmp)
gc(full=TRUE) # RES = 1771M
Garbage collection 12 = 1+1+10 (level 2) ... 
1457.1 Mbytes of cons cells used (62%)
370.6 Mbytes of vectors used (73%)
           used   (Mb) gc trigger   (Mb) limit (Mb) max used   (Mb)
Ncells 27281915 1457.1   43734150 2335.7         NA 27281989 1457.1
Vcells 48568926  370.6   66784272  509.6      16384 48570044  370.6


loaded <- readRDS(tmp)
gc(full=TRUE, verbose=TRUE) # RES = 3459M
Garbage collection 16 = 2+1+13 (level 2) ... 
1457.1 Mbytes of cons cells used (39%)
370.6 Mbytes of vectors used (48%)
           used   (Mb) gc trigger   (Mb) limit (Mb) max used   (Mb)
Ncells 27281840 1457.1   70585438 3769.7         NA 54281908 2899.0
Vcells 48568801  370.6  100774347  768.9      16384 93568945  713.9
```

~What's going on here is you can see that the amount of memory used stays constant, **but the fraction changes by a factor of 2**. That fraction (I believe) represents R's heap.~

Since R manages it's own memory it has it's own pool (heap). Garbage collection of a removed object does not necessarily change R's heap size, it just frees up the cells in the heap for use for other R objects.~ 

Hope that helps!

**Reply by wlandau** [reply]
Thanks for explaining.

How good would you say are lobster::obj_size() and lobster::mem_used() for tracking the space consumed by the in-memory R objects we can control as R users? I have been using these functions to test and verify that {targets} unloads the right objects at the proper time.

**Comment by traversc** [comment]
lobstr::mem_used is a wrapper around gc, so it should be the same. 

Reading more of the docs, I need to walk back my previous conclusion. The percentage is telling you when gc will automatically trigger. So now I'm not sure why htop res is appearing higher in these examples. 😂

It might be good to raise this question (using base R) to someone more knowledgeable on R's memory handling.

**Comment by Tristan-Siegfried** [comment]
@wlandau just wanted to follow up on this to clarify: In light of the above comment I was wondering whether you were planning to look further into this memory usage, or whether this not being an explicitly `targets` issue means it is out of scope and won't be addressed further here?

**Reply by wlandau** [reply]
Given @traversc's edifying comments, I suspect your findings may have to do with R's own memory heap, which is outside the scope of `targets`. I have tests which use `lobstr::mem_used()` and `lobstr::obj_size()` to monitor the memory consumed by the objects `targets` is responsible for, and in those tests memory usage appears as expected. So I am not sure what action I can take.
